akka-http-core/src/main/scala/akka/http/impl/engine/ws/Websocket.scala

```diff
@@ -68,7 +68,7 @@ private[http] object Websocket {
       var inMessage = false
       def onPush(elem: MessagePart, ctx: Context[MessagePart]): SyncDirective = elem match {
         case PeerClosed(code, reason) ⇒
-          if (code.exists(Protocol.CloseCodes.isError)) ctx.fail(new ProtocolException(s"Peer closed connection with code $code"))
+          if (code.exists(Protocol.CloseCodes.isError)) ctx.fail(new PeerClosedConnectionException(code.get, reason))
           else if (inMessage) ctx.fail(new ProtocolException(s"Truncated message, peer closed connection in the middle of message."))
           else ctx.finish()
         case ActivelyCloseWithCode(code, reason) ⇒
```

Comment:
Can't comment on line `.splitWhen(_.isMessageEnd) // FIXME using splitAfter from #16885 would simplify protocol a lot` https://github.com/akka/akka/pull/18750/files#diff-32111086a8e0b3bf2b812bda44d7c0a1R110 but that should be solved now https://github.com/akka/akka/issues/16885